File: src/exec.ts
import { execFileSync, spawn, type ChildProcess } from "node:child_process";
import type { Readable } from "node:stream";

import { hasErrorCode, messageOf } from "./errors.js";

/** Shape returned when the command runs to completion, whether it exits cleanly or is aborted. */
export interface ExecResult {
  stdout: string;
  stderr: string;
  exitCode: number;
  /** Total bytes received on stdout before any cap truncation. */
  stdoutBytes: number;
  /** Total bytes received on stderr before any cap truncation. */
  stderrBytes: number;
}

/** Shape returned when the command is killed for exceeding `ExecOptions.timeoutMs`. */
export interface ExecTimeoutError {
  kind: "timeout";
  stdout: string;
  stderr: string;
  timeoutMs: number;
  /** Total bytes received on stdout before any cap truncation. */
  stdoutBytes: number;
  /** Total bytes received on stderr before any cap truncation. */
  stderrBytes: number;
}

/** Options controlling where a command runs, what it reads, and how it can be stopped early. */
export interface ExecOptions {
  cwd: string;
  timeoutMs?: number;
  signal?: AbortSignal | undefined;
  /** Text written to the command's stdin, which is then closed. Omitted means no input at all. */
  stdin?: string | undefined;
}

/** Reported when the child has no exit status of its own: killed, or never started. */
export const FAILURE_EXIT_CODE = 1;

/**
 * The result for a run that never produced any output: the signal was already
 * aborted before the command could spawn, or the spawned child came up with no
 * usable stdio streams.
 */
const EMPTY_FAILURE_RESULT: ExecResult = {
  stdout: "",
  stderr: "",
  exitCode: FAILURE_EXIT_CODE,
  stdoutBytes: 0,
  stderrBytes: 0,
};

/**
 * Per-stream accumulation cap. A command that exceeds this resolves normally
 * with truncated capture rather than growing into an OOM or RangeError.
 */
export const OUTPUT_CAP_BYTES = 64 * 1024 * 1024;

/** Mutable buffer accumulating a child process's stdout/stderr as it streams in. */
interface OutputBuffer {
  stdout: string;
  stderr: string;
  stdoutBytes: number;
  stderrBytes: number;
}

/**
 * Accumulate `stdout` and `stderr` into a buffer as data arrives, so callers can
 * snapshot the output received so far at any point — on close, timeout, or abort
 * — without waiting for the streams to end.
 *
 * Each stream is independently capped at `OUTPUT_CAP_BYTES`. Once the cap is
 * reached, subsequent chunks are silently dropped — the stream stays open so the
 * child process is not signalled, but the buffer stops growing.
 */
function attachCapped(
  stream: Readable,
  buffer: OutputBuffer,
  textKey: "stdout" | "stderr",
  bytesKey: "stdoutBytes" | "stderrBytes",
): void {
  stream.on("data", (chunk: string) => {
    const chunkBytes = Buffer.byteLength(chunk, "utf8");
    buffer[bytesKey] += chunkBytes;
    if (buffer[bytesKey] - chunkBytes >= OUTPUT_CAP_BYTES) return;
    buffer[textKey] += chunk;
  });
}

function captureOutput(stdout: Readable, stderr: Readable): OutputBuffer {
  const buffer: OutputBuffer = { stdout: "", stderr: "", stdoutBytes: 0, stderrBytes: 0 };
  attachCapped(stdout, buffer, "stdout", "stdoutBytes");
  attachCapped(stderr, buffer, "stderr", "stderrBytes");
  return buffer;
}

/** Snapshot the four output fields shared by every `ExecResult`/`ExecTimeoutError` variant. */
function snapshotOutput(
  buf: OutputBuffer,
): Pick<ExecResult, "stdout" | "stderr" | "stdoutBytes" | "stderrBytes"> {
  return {
    stdout: buf.stdout,
    stderr: buf.stderr,
    stdoutBytes: buf.stdoutBytes,
    stderrBytes: buf.stderrBytes,
  };
}

/** The status taskkill exits with when no process carries the given pid. */
const TASKKILL_NOT_FOUND_STATUS = 128;

/** Kill the process group led by `pid`, descendants included. Never throws. */
function killTree(pid: number): void {
  try {
    if (process.platform === "win32") {
      // taskkill /T kills the process and all descendants; /F forces it.
      // Without this, cmd.exe dies but sh/sleep survive, holding file locks.
      execFileSync("taskkill", ["/T", "/F", "/PID", String(pid)], { stdio: "ignore" });
    } else {
      process.kill(-pid, "SIGKILL");
    }
  } catch (err) {
    // emitWarning, not throw: callers run from setTimeout/AbortSignal contexts
    // where a throw becomes an uncaught exception.
    //
    // "Process already gone" surfaces differently per platform: POSIX process.kill
    // throws with code "ESRCH", while Windows taskkill exits with
    // TASKKILL_NOT_FOUND_STATUS. Every other taskkill status is a genuine failure
    // — access denied is 5 — and has to reach the caller as a warning.
    const alreadyGone =
      process.platform === "win32"
        ? err instanceof Error && "status" in err && err.status === TASKKILL_NOT_FOUND_STATUS
        : /* istanbul ignore next -- ESRCH and EPERM need a group that dies between the kill and here */
          hasErrorCode(err, "ESRCH");
    if (!alreadyGone) {
      process.emitWarning(err instanceof Error ? err : String(err));
    }
  }
}

/**
 * The EPIPE a command that never reads its stdin leaves behind is its choice, not a fault, and
 * ERR_STREAM_DESTROYED is expected when the child exits before the write completes. Anything else
 * is a genuine failure to deliver `options.stdin` and has to reach the caller as a warning, since
 * `exec` settles from the child's own close/error events and would otherwise resolve as an
 * ordinary result with the payload silently missing.
 */
function ignoreStdinError(err: unknown): void {
  if (hasErrorCode(err, "EPIPE") || hasErrorCode(err, "ERR_STREAM_DESTROYED")) return;
  process.emitWarning(err instanceof Error ? err : String(err));
}

/**
 * Runs the command in a detached process group to enable killing all child processes
 * on timeout. On POSIX systems, sends SIGKILL to the negative PID to terminate the
 * entire process group.
 *
 * Aborting `options.signal` kills that same process group. Unlike a timeout, an abort
 * resolves with an ExecResult holding the output captured so far, so callers can tell a
 * caller-initiated cancellation apart from a timeout. A signal that is already aborted
 * when the call arrives never spawns the shell at all, and resolves with that same
 * cancelled shape.
 *
 * A run that ends on its own is snapshotted when the child's stdio closes rather than when
 * the shell exits, so a line a background job writes after the shell returns is still
 * captured. A descendant that keeps the pipe open therefore holds the call open too:
 * `timeoutMs` is what bounds it. Timeout and abort snapshot the output received so far
 * instead, so neither can be held open by a descendant that survived the group kill.
 *
 * @param command - The shell command to execute
 * @param options - Execution options (working directory, optional timeout in ms, optional
 *                  AbortSignal for cancellation, optional stdin text). The command's stdin is
 *                  closed either way, so a command that reads it sees end-of-input instead of
 *                  blocking, and one that ignores it is not held to account for the broken pipe.
 * @returns A promise resolving to either an ExecResult with the command output and exit code,
 *          or an ExecTimeoutError if the timeout is exceeded
 *
 * @example
 * const result = await exec('echo hello', { cwd: '/tmp' });
 * if ('kind' in result) {
 *   console.log('Timed out after', result.timeoutMs, 'ms');
 * } else {
 *   console.log('Output:', result.stdout);
 * }
 */
export async function exec(
  command: string,
  options: ExecOptions,
): Promise<ExecResult | ExecTimeoutError> {
  return new Promise((resolve) => {
    const spawned = trySpawn(command, options);
    if ("exitCode" in spawned) {
      resolve(spawned);
      return;
    }
    wireChildEvents(spawned, options, resolve);
  });
}

interface SpawnedChild {
  child: ChildProcess;
  stdout: Readable;
  stderr: Readable;
}

function trySpawn(command: string, options: ExecOptions): ExecResult | SpawnedChild {
  // A signal already aborted when the call arrives is a cancellation, not a run.
  if (options.signal?.aborted) {
    return EMPTY_FAILURE_RESULT;
  }

  // Some spawn failures are raised here rather than on the "error" event —
  // ENOTDIR from a cwd that is not a directory is one.
  let child: ChildProcess;
  try {
    child = spawn(command, {
      shell: true,
      cwd: options.cwd,
      // POSIX: detach into its own process group so killTree can SIGKILL
      // the whole group via negative PID. Windows has no process groups,
      // and detached: true there allocates a new console that breaks
      // stdio pipes — stdout/stderr go to the console instead of the pipe.
      detached: process.platform !== "win32",
      stdio: ["pipe", "pipe", "pipe"],
    });
  } catch (err) {
    const errMsg = `${messageOf(err)}\n`;
    return {
      stdout: "",
      stderr: errMsg,
      exitCode: FAILURE_EXIT_CODE,
      stdoutBytes: 0,
      stderrBytes: Buffer.byteLength(errMsg, "utf8"),
    };
  }

  // fd exhaustion can leave stdio as null even though spawn did not throw.
  if (child.stdout === null || child.stderr === null || child.stdin === null) {
    if (child.pid !== undefined) {
      killTree(child.pid);
    }
    child.on("error", () => {});
    return EMPTY_FAILURE_RESULT;
  }

  child.stdout.setEncoding("utf8");
  child.stderr.setEncoding("utf8");

  child.stdin.on("error", ignoreStdinError);
  child.stdin.end(options.stdin ?? "");

  return { child, stdout: child.stdout, stderr: child.stderr };
}

function wireChildEvents(
  spawned: SpawnedChild,
  options: ExecOptions,
  resolve: (value: ExecResult | ExecTimeoutError) => void,
): void {
  const { child, stdout: childStdout, stderr: childStderr } = spawned;
  const output = captureOutput(childStdout, childStderr);
  let timeoutHandle: NodeJS.Timeout | undefined;
  let resolved = false;

  function settle(result: ExecResult | ExecTimeoutError): void {
    if (resolved) return;
    resolved = true;
    clearTimeout(timeoutHandle);
    options.signal?.removeEventListener("abort", onAbort);
    childStdout.destroy();
    childStderr.destroy();
    // oxlint-disable-next-line promise/no-multiple-resolved
    resolve(result);
  }

  function killGroup(): void {
    if (!child.pid) return;
    killTree(child.pid);
  }

  function onAbort(): void {
    killGroup();
    settle({ ...snapshotOutput(output), exitCode: FAILURE_EXIT_CODE });
  }

  const { timeoutMs } = options;
  if (timeoutMs) {
    timeoutHandle = setTimeout(() => {
      killGroup();
      settle({ kind: "timeout", ...snapshotOutput(output), timeoutMs });
    }, timeoutMs);
  }

  options.signal?.addEventListener("abort", onAbort, { once: true });

  child.on("close", (code) => {
    settle({ ...snapshotOutput(output), exitCode: code ?? FAILURE_EXIT_CODE });
  });

  function onFailure(err: Error): void {
    killGroup();
    const errSuffix = `${err.message}\n`;
    settle({
      ...snapshotOutput(output),
      stderr: `${output.stderr}${errSuffix}`,
      exitCode: FAILURE_EXIT_CODE,
      stderrBytes: output.stderrBytes + Buffer.byteLength(errSuffix, "utf8"),
    });
  }

  child.on("error", onFailure);
  childStdout.on("error", onFailure);
  childStderr.on("error", onFailure);
}

File: src/gymrat_py/exec.py
"""Run shell commands as asyncio subprocesses with bounded, decoded capture.

``exec`` spawns a command through the shell, captures stdout and stderr
separately as decoded text plus raw byte counts, and settles when the child's
stdio pipes close — not merely when the shell exits, so output flushed by a
background descendant is still captured. A run is bounded three ways:

- a per-stream 64 MiB text cap (byte counts keep counting past it),
- an optional timeout that resolves an :class:`ExecTimeoutError` value, and
- an optional abort :class:`asyncio.Event` that resolves a failed
  :class:`ExecResult`.

Timeout and abort snapshot whatever has been captured so far and kill the whole
process group, so a grandchild the shell left running is reaped too. Every
returned value is a frozen dataclass, so a caller cannot mutate one run's result
into a landmine for the next.
"""

import asyncio
import codecs
import contextlib
import os
import signal
import subprocess
import sys
import warnings
from collections.abc import Mapping
from dataclasses import dataclass

FAILURE_EXIT_CODE = 1
"""Exit code reported when a run fails without a positive child exit code."""

OUTPUT_CAP = 64 * 1024 * 1024
"""Per-stream cap, in bytes, on retained text. Byte counts keep counting past it."""

_READ_CHUNK = 65536
"""Bytes requested per pipe read."""

_TASKKILL_GONE = 128
"""``taskkill`` exit status meaning the process was already gone."""


@dataclass(frozen=True, slots=True)
class ExecOptions:
    """Inputs for a single :func:`exec` run.

    Args:
        cwd: Working directory the command runs in.
        timeout_ms: Wall-clock budget in milliseconds; ``None`` waits forever.
        abort: Event that, once set, kills the run and resolves a failed result.
        stdin: Text delivered to the command's standard input, then closed;
            ``None`` gives an immediately closed (EOF) input.
        env: Variables merged *over* the inherited environment; ``None`` inherits
            the environment unchanged.
    """

    cwd: str
    timeout_ms: int | None = None
    abort: asyncio.Event | None = None
    stdin: str | None = None
    env: Mapping[str, str] | None = None


@dataclass(frozen=True, slots=True)
class ExecResult:
    """A completed run's captured output and exit code."""

    stdout: str
    stderr: str
    exit_code: int
    stdout_bytes: int
    stderr_bytes: int


@dataclass(frozen=True, slots=True)
class ExecTimeoutError:
    """A run that exceeded its timeout, with whatever was captured beforehand.

    This is a returned value, not a raised exception: callers distinguish it from
    :class:`ExecResult` by type (``isinstance``).
    """

    stdout: str
    stderr: str
    timeout_ms: int
    stdout_bytes: int
    stderr_bytes: int


@dataclass(slots=True)
class OutputBuffer:
    """Accumulates decoded text up to :data:`OUTPUT_CAP` while counting all bytes.

    A chunk is appended only while the bytes received *before* it are still under
    the cap; the chunk that crosses the cap is kept whole, and every later chunk
    is dropped from the text. ``byte_count`` always reflects every byte received,
    so a caller can tell that output was truncated.
    """

    text: str = ""
    byte_count: int = 0

    def append(self, chunk: str, chunk_bytes: int) -> None:
        """Add ``chunk`` (worth ``chunk_bytes`` raw bytes) subject to the cap."""
        self.byte_count += chunk_bytes
        if self.byte_count - chunk_bytes >= OUTPUT_CAP:
            return
        self.text += chunk

    def append_failure(self, message: str) -> None:
        """Append a failure ``message`` and newline unconditionally, past the cap.

        An error explaining why a run failed must always reach the caller, so it
        bypasses the truncation cap that governs ordinary output.
        """
        line = f"{message}\n"
        self.text += line
        self.byte_count += len(line.encode())


def _platform() -> str:
    """Return the current platform.

    Read at kill time rather than cached at spawn so tests can redirect the kill
    strategy to the Windows path after a POSIX spawn.
    """
    return sys.platform


def _exit_code(returncode: int | None) -> int:
    """Map a child's raw return code to a reported exit code.

    A signal kill surfaces as a negative return code, and a child whose status
    has not been collected yet as ``None``; both collapse to
    :data:`FAILURE_EXIT_CODE` rather than leaking a negative or missing value.
    """
    if returncode is None or returncode < 0:
        return FAILURE_EXIT_CODE
    return returncode


def _taskkill(pid: int) -> None:
    """Run ``taskkill /T /F`` for ``pid``, raising on any non-zero status.

    ``taskkill`` resolves from ``PATH`` on Windows and the argv is fixed, so the
    subprocess-injection check does not apply.
    """
    argv = ["taskkill", "/F", "/T", "/PID", str(pid)]
    subprocess.run(argv, capture_output=True, check=True)  # noqa: S603


def _kill_group(pid: int) -> None:
    """Kill the whole process group led by ``pid``, never raising into the caller.

    On POSIX this signals the group with ``SIGKILL``, staying silent when the
    group is already gone and warning on any other failure. On Windows it
    delegates to ``taskkill /T /F``, staying silent when the process is already
    gone and warning on any other failure.
    """
    if _platform() == "win32":
        try:
            _taskkill(pid)
        except subprocess.CalledProcessError as error:
            if error.returncode != _TASKKILL_GONE:
                warnings.warn(
                    f"taskkill failed for pid {pid}: {error}",
                    RuntimeWarning,
                    stacklevel=2,
                )
        except FileNotFoundError:
            warnings.warn(
                f"taskkill unavailable while killing pid {pid}",
                RuntimeWarning,
                stacklevel=2,
            )
        return
    try:
        os.killpg(pid, signal.SIGKILL)
    except ProcessLookupError:
        pass
    except OSError as error:
        warnings.warn(
            f"killpg failed for pid {pid}: {error}",
            RuntimeWarning,
            stacklevel=2,
        )


def _terminate(proc: asyncio.subprocess.Process) -> None:
    """Kill ``proc``'s process group and drop its stdio pipes."""
    _kill_group(proc.pid)
    _close_pipes(proc)


def _close_pipes(proc: asyncio.subprocess.Process) -> None:
    """Close the child's stdio pipes so a surviving descendant cannot grow buffers.

    Closing the transport drops the read ends, and feeding EOF to each reader
    settles them deterministically for a snapshot taken without waiting for the
    natural end of stream.
    """
    # asyncio.subprocess.Process exposes no public accessor for its pipe
    # transports; reaching the undocumented transport via getattr is the only way
    # to drop the read ends before the child is reaped.
    transport = getattr(proc, "_transport", None)
    if transport is not None:
        transport.close()
    for reader in (proc.stdout, proc.stderr):
        if reader is not None and not reader.at_eof():
            reader.feed_eof()


async def _read_stream(reader: asyncio.StreamReader, buffer: OutputBuffer) -> None:
    """Read ``reader`` to EOF, decoding UTF-8 incrementally into ``buffer``.

    One incremental decoder per stream reassembles a multi-byte character split
    across pipe reads; the raw byte length of each read drives the byte counts.
    """
    decoder = codecs.getincrementaldecoder("utf-8")(errors="replace")
    while True:
        chunk = await reader.read(_READ_CHUNK)
        if not chunk:
            tail = decoder.decode(b"", final=True)
            if tail:
                buffer.append(tail, 0)
            return
        buffer.append(decoder.decode(chunk), len(chunk))


async def _feed_stdin(proc: asyncio.subprocess.Process, data: str | None) -> None:
    """Write ``data`` to the child's stdin then close it, swallowing broken pipes.

    A child that exits without reading a large payload breaks the write; that is
    an expected end-of-run condition, not an error to surface.
    """
    stdin = proc.stdin
    if stdin is None:
        return
    try:
        if data:
            stdin.write(data.encode())
            await stdin.drain()
    except (BrokenPipeError, ConnectionResetError):
        pass
    finally:
        with contextlib.suppress(BrokenPipeError, ConnectionResetError, OSError):
            stdin.close()


async def _await_normal(
    proc: asyncio.subprocess.Process,
    stdout_task: asyncio.Task[None],
    stderr_task: asyncio.Task[None],
) -> None:
    """Wait for both readers to reach EOF and the child to be reaped.

    Settling on stdio close rather than process exit is what captures output a
    background descendant flushes after the shell itself has returned.
    """
    await asyncio.gather(stdout_task, stderr_task)
    await proc.wait()


async def _cancel_all(tasks: list[asyncio.Task[object]]) -> None:
    """Cancel every task and await their settling, discarding their outcomes."""
    for task in tasks:
        if not task.done():
            task.cancel()
    await asyncio.gather(*tasks, return_exceptions=True)


def _build_result(stdout_buf: OutputBuffer, stderr_buf: OutputBuffer, exit_code: int) -> ExecResult:
    """Assemble an :class:`ExecResult` from captured buffers and an exit code."""
    return ExecResult(
        stdout_buf.text,
        stderr_buf.text,
        exit_code,
        stdout_buf.byte_count,
        stderr_buf.byte_count,
    )


async def exec(command: str, options: ExecOptions) -> ExecResult | ExecTimeoutError:  # noqa: A001
    """Run ``command`` through the shell and capture its output.

    The run never raises for a spawn or child failure: a shell that cannot be
    spawned, a non-zero exit, a signal kill, an abort, and a stream error all
    resolve as an :class:`ExecResult`. Only exceeding ``options.timeout_ms``
    resolves the distinct :class:`ExecTimeoutError` value.

    Args:
        command: The shell command line to run.
        options: Working directory, timeout, abort event, stdin, and environment.

    Returns:
        An :class:`ExecResult` for any completed or cancelled run, or an
        :class:`ExecTimeoutError` when the timeout is exceeded.
    """
    if options.abort is not None and options.abort.is_set():
        return ExecResult("", "", FAILURE_EXIT_CODE, 0, 0)

    env = {**os.environ, **options.env} if options.env is not None else None
    try:
        proc = await asyncio.create_subprocess_shell(
            command,
            cwd=options.cwd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=env,
            start_new_session=_platform() != "win32",
        )
    except OSError as error:
        stderr = f"{error}\n"
        return ExecResult("", stderr, FAILURE_EXIT_CODE, 0, len(stderr.encode()))

    stdout_buf = OutputBuffer()
    stderr_buf = OutputBuffer()
    assert proc.stdout is not None  # noqa: S101 -- guaranteed by stdout=PIPE
    assert proc.stderr is not None  # noqa: S101 -- guaranteed by stderr=PIPE
    stdout_task = asyncio.create_task(_read_stream(proc.stdout, stdout_buf))
    stderr_task = asyncio.create_task(_read_stream(proc.stderr, stderr_buf))
    stdin_task = asyncio.create_task(_feed_stdin(proc, options.stdin))
    normal_task = asyncio.create_task(_await_normal(proc, stdout_task, stderr_task))
    abort_task = asyncio.create_task(options.abort.wait()) if options.abort is not None else None

    waiters: list[asyncio.Task[object]] = [normal_task]
    if abort_task is not None:
        waiters.append(abort_task)
    timeout = options.timeout_ms / 1000 if options.timeout_ms is not None else None

    try:
        done, _ = await asyncio.wait(
            waiters,
            timeout=timeout,
            return_when=asyncio.FIRST_COMPLETED,
        )

        if normal_task in done:
            reader_error = normal_task.exception()
            if reader_error is not None:
                _terminate(proc)
                stderr_buf.append_failure(str(reader_error))
                return _build_result(stdout_buf, stderr_buf, FAILURE_EXIT_CODE)
            return _build_result(stdout_buf, stderr_buf, _exit_code(proc.returncode))

        _terminate(proc)
        if abort_task is not None and abort_task in done:
            return _build_result(stdout_buf, stderr_buf, FAILURE_EXIT_CODE)
        if options.timeout_ms is not None:
            return ExecTimeoutError(
                stdout_buf.text,
                stderr_buf.text,
                options.timeout_ms,
                stdout_buf.byte_count,
                stderr_buf.byte_count,
            )
        # asyncio.wait only returns empty-handed when a timeout elapsed, so with
        # no timeout set this branch cannot be reached.
        msg = "exec settled without a normal, abort, or timeout outcome"
        raise RuntimeError(msg)
    finally:
        pending: list[asyncio.Task[object]] = [
            stdout_task,
            stderr_task,
            stdin_task,
            normal_task,
        ]
        if abort_task is not None:
            pending.append(abort_task)
        await _cancel_all(pending)
